compiler/rustc_abi/src/layout.rs

```diff
@@ -897,42 +897,58 @@ pub trait LayoutCalculator {
             align = align.max(AbiAndPrefAlign::new(repr_align));
         }
 
-        let optimize = !repr.inhibit_union_abi_opt();
+        let mut optimize = !repr.inhibit_union_abi_opt();
         let mut size = Size::ZERO;
-        let mut abi = Abi::Aggregate { sized: true };
+        let mut common_non_zst_abi_and_align: Option<(Abi, AbiAndPrefAlign)> = None;
         let index = V::new(0);
         for field in &variants[index] {
             assert!(field.is_sized());
-            align = align.max(field.align);
 
-            // If all non-ZST fields have the same ABI, forward this ABI
-            if optimize && !field.is_zst() {
+            if !field.is_zst() && optimize {
                 // Discard valid range information and allow undef
-                let field_abi = match field.abi {
-                    Abi::Scalar(x) => Abi::Scalar(x.to_union()),
-                    Abi::ScalarPair(x, y) => Abi::ScalarPair(x.to_union(), y.to_union()),
-                    Abi::Vector { element: x, count } => {
-                        Abi::Vector { element: x.to_union(), count }
+                let field_abi = field.abi.to_union();
+
+                if let Some((abi, align)) = &mut common_non_zst_abi_and_align {
+                    if *abi != field_abi {
+                        // Different fields have different ABI: disable opt
+                        optimize = false;
```

Comment:
Might be better to have a `common_non_zst_abi` be a `Result<Option<Abi>, AbiMismatch>`?
(with a locally defined `struct AbiMismatch;` - at least that's the kind of thing I tend to do to label those kinds of situations, I think?)